pandas/io/orc.py

```diff
@@ -52,3 +59,78 @@ def read_orc(
+
+
+def to_orc(
+    df: DataFrame,
+    path: FilePath | WriteBuffer[bytes] | None = None,
+    engine: Literal['pyarrow'] = 'pyarrow',  # type: ignore[arg-type]
+    index: bool = None,
+    **kwargs
+) -> bytes:
+    """
+    Write a DataFrame to the ORC format.
+    Parameters
+    ----------
+    df : DataFrame
+    path : str or file-like object, default None
+        If a string, it will be used as Root Directory path
+        when writing a partitioned dataset. By file-like object,
+        we refer to objects with a write() method, such as a file handle
+        (e.g. via builtin open function). If path is None,
+        a bytes object is returned. Note that currently the pyarrow
+        engine doesn't work with io.BytesIO.
+    engine : {{'pyarrow'}}, default 'pyarrow'
+        Parquet library to use, or library it self, checked with 'pyarrow' name
+        and version >= 5.0.0
+    index : bool, default None
+        If ``True``, include the dataframe's index(es) in the file output. If
+        ``False``, they will not be written to the file.
+        If ``None``, similar to ``infer`` the dataframe's index(es)
+        will be saved. However, instead of being saved as values,
+        the RangeIndex will be stored as a range in the metadata so it
+        doesn't require much space and is faster. Other indexes will
+        be included as columns in the file output.
+    kwargs
+        Additional keyword arguments passed to the engine
+    Returns
+    -------
+    bytes if no path argument is provided else None
+    """
+    if index is None:
+        index = df.index.names[0] is not None
+
+    if isinstance(engine, str):
+        engine = import_optional_dependency(engine, min_version='5.0.0')
+    elif isinstance(engine, ModuleType):
+        try:
+            assert engine.__name__ == 'pyarrow', "engine must be 'pyarrow' module"
+            assert hasattr(engine, 'orc'), "'pyarrow' module must have orc module"
+        except ImportError as e:
+            raise ValueError(
+                "Unable to find a usable engine; "
+                "tried using: 'pyarrow'.\n"
+                "A suitable version of "
+                "pyarrow is required for ORC support.\n"
+                "Trying to import the above resulted in these errors:"
+                f"\n - {e}"
+            )
+    else:
+        raise TypeError(
+            f"unsuported type for engine: {type(engine)}"
+        )
+
+    if hasattr(path, "write"):
+        engine.orc.write_table(
+            engine.Table.from_pandas(df, preserve_index=index),
+            path, **kwargs
+        )
+    else:
+        # to bytes: pyarrow auto closes buffers hence we read a pyarrow buffer
+        stream = engine.BufferOutputStream()
+        engine.orc.write_table(
+            engine.Table.from_pandas(df, preserve_index=index),
+            stream, **kwargs
+        )
+        return stream.getvalue().to_pybytes()
```

Comment:
I think this if/else could be:

```py
if not hasattr(path, "write"):
    ...
else:
    with engine.BufferOutputStream() as stream:  # if that is possible
        engine.orc.write_table(...)
        with get_handle(path, "wb", is_text=False) as handles:  # allows writing to any (fsspec) URL
            handles.handle.write(stream.getvalue().to_pybytes())
```

Reply:
I made it the following in order to make sure we actually return the ORC bytes if `path` is `None`
      if hasattr(path, "write"):
          engine.orc.write_table(
              engine.Table.from_pandas(df, preserve_index=index),
              path, **kwargs
          )
      else:
          # to bytes: pyarrow auto closes buffers hence we read a pyarrow buffer
          with engine.BufferOutputStream() as stream:  # if that is possible
              engine.orc.write_table(
                  engine.Table.from_pandas(df, preserve_index=index),
                  stream, **kwargs
              )
              # allows writing to any (fsspec) URL
              with get_handle(path, "wb", is_text=False) as handles:
                  orc_bytes = stream.getvalue().to_pybytes()
                  handles.handle.write(orc_bytes)
                  if path is None:
                      return orc_bytes

Reply:
Looks good! Small changes (re-arranging some lines):

```py
if hasattr(path, "write"):
    engine.orc.write_table(
        engine.Table.from_pandas(df, preserve_index=index),
        path, **kwargs
    )
else:
    # to bytes: pyarrow auto closes buffers hence we read a pyarrow buffer
    with engine.BufferOutputStream() as stream:  # if that is possible
        engine.orc.write_table(
            engine.Table.from_pandas(df, preserve_index=index),
            stream, **kwargs
        )
        orc_bytes = stream.getvalue().to_pybytes()
        if path is None:
            return orc_bytes
        # allows writing to any (fsspec) URL
        with get_handle(path, "wb", is_text=False) as handles:
            handles.handle.write(orc_bytes)
```

Reply:
Thanks again! Done! Before adding roundtrip tests I may need to make some changes on the Arrow side.